std/os/linux/tls.zig

```diff
@@ -0,0 +1,251 @@
+};
+
+// Controls how many bytes are reserved for the Thread Control Block
+const tls_tcb_size = switch (builtin.arch) {
+    // ARM EABI mandates enough space for two pointers: the first one points to
+    // the DTV while the second one is unspecified but reserved
+    .arm, .armeb, .aarch64, .aarch64_be => 2 * @sizeOf(usize),
+    .i386, .x86_64 => @sizeOf(usize),
+    else => 0,
+};
+
+// Controls if the TCB should be aligned according to the TLS segment p_align
+const tls_tcb_align_size = switch (builtin.arch) {
+    .arm, .armeb, .aarch64, .aarch64_be => true,
+    else => false,
+};
+
+// Check if the architecture-specific parameters look correct
+comptime {
+    if (tls_tcb_align_size and tls_variant != TLSVariant.VariantI) {
+        @compileError("tls_tcb_align_size is only meaningful for variant I TLS");
+    }
+}
+
+// Some architectures add some offset to the tp and dtv addresses in order to
+// make the generated code more efficient
+
+const tls_tp_offset = switch (builtin.arch) {
+    else => 0,
+};
+
+const tls_dtv_offset = switch (builtin.arch) {
+    else => 0,
+};
+
+// Per-thread storage for Zig's use
+const CustomData = packed struct {
+};
+
+// Dynamic Thread Vector
+const DTV = packed struct {
+    entries: usize,
+    tls_block: [1]usize,
+};
+
+// Holds all the information about the process TLS image
+const TLSImage = struct {
+    data_src: []u8,
+    alloc_size: usize,
+    tcb_offset: usize,
+    dtv_offset: usize,
+    data_offset: usize,
+};
+
+pub var tls_image: ?TLSImage = null;
+
+pub fn setThreadPointer(addr: usize) void {
+    switch (builtin.arch) {
+        .x86_64 => {
+            const ARCH_SET_FS = 0x1002;
+            const rc = std.os.linux.syscall2(std.os.linux.SYS_arch_prctl, ARCH_SET_FS, addr);
+            // arch_prctl is documented to never fail
+            assert(rc == 0);
+        },
+        .aarch64 => {
+            asm volatile (
+                \\ msr tpidr_el0, %[addr]
+                : : [addr] "r" (addr)
+            );
+        },
+        else => @compileError("Unsupported architecture"),
+    }
+}
+
+pub fn initTLS() void {
+    var tls_phdr: ?*elf.Phdr = null;
+    var img_base: usize = 0;
+
+    if (std.os.linux_elf_aux_maybe) |auxv| {
+        var at_phent: usize = undefined;
+        var at_phnum: usize = undefined;
+        var at_phdr: usize = undefined;
+
+        var i: usize = 0;
+        while (auxv[i].a_type != std.elf.AT_NULL) : (i += 1) {
+            switch (auxv[i].a_type) {
+                elf.AT_PHENT => at_phent = auxv[i].a_un.a_val,
+                elf.AT_PHNUM => at_phnum = auxv[i].a_un.a_val,
+                elf.AT_PHDR  => at_phdr  = auxv[i].a_un.a_val,
+                else => continue,
+            }
+        }
+
+        // Sanity check
+        assert(at_phent == @sizeOf(elf.Phdr));
+
+        // Search the TLS section
+        const phdrs = (@intToPtr([*]elf.Phdr, at_phdr))[0..at_phnum];
+
+        for (phdrs) |*phdr| {
```

Comment:
What should happen if e.g. `PT_PHDR` is encountered twice?

Reply:
If that happens your ELF file is hosed and all bets are off.

Reply:
I feel like we should have an `assert` for it...